internal/edittree/edittree.go

```diff
@@ -0,0 +1,1135 @@
+// to rebuild the whole Term, sans the modified/deleted bits.
+//
+// To work around that problem, the EditTree allows simple add, modify, or
+// delete operations on Term structures, and then at the end of a series of
+// edits, the caller can pay the cost of generating a new Term from the
+// tree of edits relatively efficiently. (Essentially a recursive, DFS
+// traversal of the tree.)
+//
+// The data structure preserves basic type/safety properties, the same as
+// working on the real underlying Term values. To do this, recursive
+// lookups are used. On average, these are fairly straightforward and
+// cheap to do.
+//
+// Basic Operations:
+//  - Insert/Update EditTree node
+//  - Delete EditTree node
+//  - Unfold EditTree nodes along a path
+//  - Render EditTree node
+//
+// These operations provide all of the basic utilities required to
+// recursively construct the tree. Ref-based convenience functions are also
+// provided, to make rendering subtrees at a particular JSON path easier.
+//
+// Path-Based Convenience Functions:
+//  - InsertAtPath
+//  - DeleteAtPath
+//  - RenderAtPath
+//
+// Additionally, a few "optional" (but nice to have) functions have been
+// added, to allow replacing slower/less-efficient equivalents elsewhere.
+//
+// Optional Functions:
+//  - Exists (a more efficient boolean alternative to Unfold)
+//  - Filter (an alternative to Object.Filter that efficiently renders
+//    paths out of an EditTree)
+//
+// --- Storing scalar children "inline" ---
+// The original design for the EditTree allocated a new EditTree node for
+// each Term stored in the tree, but this was found to be inefficient when
+// dealing with large arrays and objects. The current design of the
+// EditTree separates children based on their types, with scalars stored in
+// a hash -> Term map, and composites stored in a hash -> EditTree map.
+//
+// This results in dramatically fewer heap allocations and faster access
+// times for "shallow" Term structures, without penalizing nested Term
+// structures noticeably.
+//
+// --- Object operations ---
+// Objects are the most straightforward composite type, as their key-value
+// structure maps naturally onto trees. Their inserts/deletes are recorded
+// directly in the appropriate child maps, with almost no additional
+// complexity required.
+//
+// --- Set operations ---
+// Set data types have a major problem: they're *content-addressed*. This
+// means that we often have to render/materialize the sub-terms before
+// carrying out inserts or deletes, in order to know if the path to the
+// destination Term exists. This forces a tree collapse at the Set's
+// EditTree node, and is brutally inefficient.
+//
+// Example:
+//  Source set: {[0], "a"}
+//  {"op": "add", "path": [[0], 1], "value": 1}      -> result value: {[0, 1], "a"}
+//  {"op": "add", "path": [[0, 1], 2], "value": 2}   -> result value: {[0, 1, 2], "a"}
```

Comment:
[nit] It's not really ambiguous for more than few seconds, but let's use different numbers anyways? 😄 
```suggestion
//  {"op": "add", "path": [[0, 1], 2], "value": 3}   -> result value: {[0, 1, 3], "a"}
```

Reply:
Absolutely! I've rolled in your suggestion to my latest changes. :+1: